### PR TITLE
Add a function to test if an arbitrary address is a valid object reference

### PR DESCRIPTION
**Pull request by wks**
This PR adds a function to support conservative GC.  It tests whether a given address (any address, no requirement) points to an object allocated in MMTk.

The function is named `memory_manager::is_mmtk_object`.  The existing `SFT::is_mmtk_object` is renamed to `SFT::is_in_space`, and `memory_manager::is_mapped_object is renamed to is_in_mmtk_spaces`.

This function requires the `global_alloc_bit` feature.

This PR does not change `memory_manager::is_mapped_address`.  That function is used only in JikesRVM for two purposes: testing if an object is in the heap, and testing if the code is (JIT-compiled) VM code, not native code. I am still uncertain about the appropriate approach for JikesRVM + Rust MMTk.

### CODE REVIEW

File: .github/scripts/ci-test.sh
Comment (qinsoon) [inline .github/scripts/ci-test.sh:21]: This makes sure no regression for semispace and marksweep on those tests. But it does not prevent other plans and new plans breaking the tests. 

It would be good if we could extract plans from `PlanSelector`, like this in Python: https://github.com/mmtk/mmtk-core/blob/e85d99761935da867d766cec48e1fcdd941e8286/examples/build.py#L18
and run the tests for all the plans and all the features (some tests will be skipped if the feature they require is not enabled). Like what we do for `cargo test`.

No change is requested on this for the PR.

**Review by qinsoon — CHANGES_REQUESTED** [review]
Good pull request. It clarifies some ambiguity that we have for a long time. However, there are two issues that I feel need to be addressed before this PR is approved:
1. These new methods should work correctly with `vm_space`. Their documentation/semantic needs to take `vm_space` into consideration. 
2. I suggest not differentiating addresses that are mapped to the same bit in side metadata in those checks. As we use 1 alloc bit per 8 bytes, it is impossible to tell which of those 8 bytes actually has the alloc bit. We can use some heuristics to filter and guess, but I don't think it's worth it.

File: src/memory_manager.rs
Comment (wks) [reply src/memory_manager.rs:335]: I am more biased towards considering `vm_space` as "in MMTk spaces", like other spaces.  If the VM allocates the memory of the `vm_space` and uses it, it must know its range, and can tell if an object is in the `vm_space` by its address range.

I'll update the comments to clarify the results for the `vm_space`.

File: src/policy/space.rs
Comment (wks) [reply src/policy/space.rs:333]: We need to be precise here.  Returning `true` means the GC will attempt to trace through it. It will try to access object fields using `addr.to_object_reference()` as the object reference.  If `addr` is slightly misaligned, the GC will end up reading fields at wrong offsets, too.

I'll change the code and take the offset into consideration.

File: .github/scripts/ci-test.sh
Comment (wks) [reply .github/scripts/ci-test.sh:21]: MarkSweep and SemiSpaces are just two representatives for plans that has external allocators (MallocSpace) and the plans that does not have such allocators.  But it is always a good idea to test it for all plans.

**Comment by wks** [comment]
binding-refs
OPENJDK_BINDING_REPO=wks/mmtk-openjdk
OPENJDK_BINDING_REF=is_arbitrary_address_alloced
JIKESRVM_BINDING_REPO=wks/mmtk-jikesrvm
JIKESRVM_BINDING_REF=is_arbitrary_address_alloced
V8_BINDING_REPO=wks/mmtk-v8
V8_BINDING_REF=is_arbitrary_address_alloced

File: src/vm/object_model.rs
Comment (qinsoon) [inline src/vm/object_model.rs:104]: This is a very strong requirement. This essentially requires that all the object references starts at the same byte in a word. It is very likely to be true for a binding, but we cannot guarantee that's always true for all the VMs.

For 64 bits, our default min alignment is 4 bytes and min object size is 8 bytes. That means a binding may get addresses like `0x0`, `0x14` from allocation. If the binding uses the addresses as object refs, those two object refs do not have the same `OBJREF_ALLOC_BIT_REM`.

I think this should be left to the binding to check. That would save us lots of trouble to make sure it is general and correct, and for each specific binding, it is very easy for them to filter out obviously invalid references based on the information they have.

File: vmbindings/dummyvm/Cargo.toml
Comment (qinsoon) [inline vmbindings/dummyvm/Cargo.toml:5]: Any specific reason to add edition 2021 here? `mmtk-core` is using edition `2018` (though we should migrate to 2021).

**Review by qinsoon — CHANGES_REQUESTED** [review]

File: src/memory_manager.rs
Comment (wks) [reply src/memory_manager.rs:315]: Good idea. I'll just mention there are special behaviours when there is VM space and link to the VM space documentation.

File: src/memory_manager.rs
Comment (wks) [reply src/memory_manager.rs:365]: Yes. That's reasonable.

I introduced both because I think the semantics of `is_mmtk_object_prechecked` is a bit hard to express without the plain `is_mmtk_object`.  But from your comments below, I think we'd better leave it to the VM.  We will still require the VM to guarantee that there cannot be two objects competing the same alloc bit.

File: vmbindings/dummyvm/Cargo.toml
Comment (wks) [reply vmbindings/dummyvm/Cargo.toml:5]: I used the logging facilities (`info!(...)`, `debug!(...)`, ...) in dummyvm and found the macros need to be imported differently than `mmtk-core`.  [The rules for importing macros changed in 2018 edition.](https://doc.rust-lang.org/edition-guide/rust-2018/path-changes.html#macros)  I had two options:

1.  Keep it as 2015, and add `use` statements in the 2015 style.
2.  Add `edition=2018` or `edition=2021` to `Cargo.toml`.

I think we will eventually migrate to a new edition, so there is pointless to stick to 2015.  For 2018 vs 2021, Rust is designed so that [changing the edition of one crate does not affect another crate](https://doc.rust-lang.org/edition-guide/editions/index.html#editions-do-not-split-the-ecosystem).  So I skipped 2018 and set it to 2021 directly.

File: src/vm/object_model.rs
Comment (wks) [reply src/vm/object_model.rs:104]: The alignment requirement is `addr % MIN_ALIGN == OBJREF_ALLOC_BIT_REM`. It is `% MIN_ALIGN`, not `% MIN_SIZE`. In this case, `0x0` and `0x14` are both congruent to 0 modulo 4 (not modulo 8).

The requirement that "all valid objrefs must have the same remainder" is strong, but I think it has to be this strong.  

Remember that [JikesRVM finds the header size of a valid objref by loading from the object header](https://github.com/mmtk/jikesrvm/blob/master/rvm/src/org/jikesrvm/objectmodel/JavaHeader.java#L328).  When we are given an address to check if it is a valid ObjectReference, it can be an arbitrary address.  It is not safe to load from an arbitrary address, not even when it is in one of MMTk spaces.

`addr % MIN_ALIGN` can be 0, 1, 2, or 3 if `MIN_ALIGN == 4`.  If different objrefs may have different remainders, it is impossible to tell which is the proper remainder.  We can't look at object headers because we can't load from arbitrary addresses.  Then we can only find the proper remainder from other data sources, such as address ranges and side metadata.  But I don't think either of them is practical.

But anyway, we can offload this to the VM, because in practice, the VM will do the pre-check anyway.

File: src/policy/space.rs
Comment (wks) [reply src/policy/space.rs:95]: When I added the assertion, it failed when `addr = actual_objref + 4194304` in the `tests::conservatism::large_offsets_aligned` test case.  I am trying to find why.

Comment (wks) [reply src/policy/space.rs:95]: The reason is, SFT is allocated in chunks, and a chunk is 4MB.  In my testing fixture, I set the heap size to only 2MB. With SemiSpace, each CopySpace occupies 1MB.  There are still 2MB in the chunk not mapped to anything. So `actual_objref + 4194304` points to the same SFT chunk, but the `addr` itself is not mapped.

Comment (qinsoon) [reply src/policy/space.rs:95]: Interesting... Probably just ignore my comment on this. I will create an issue and look at this later. If as you said each copy space is 1MB, they probably use the same SFT entry which seems wrong. I am wondering why there isn't an assertion failure somewhere.

Comment (wks) [reply src/policy/space.rs:95]: After some debugging, I found that SFT entries for contiguous spaces are initialised eagerly during `gc_init`.  When using SemiSpace, SFT will eagerly map each chunk of the entire address range (4TB) of the copyspace0 to it, so SFT_MAP will always call `SFT::is_mmtk_object` as long as the `addr` is within the 4TB range.  However, actual memory chunks are mmap-ed on demand, and a chunk is 4MB (`const LOG_BYTES_IN_CHUNK: usize = 22`). This is why the assertion fails at 4MB offset.

So the check for `addr.is_mapped()` is still necessary.

**Review by qinsoon — APPROVED** [review]
LGTM

**Pull request merged**
